Provide separate functions for creating tx payloads, and fulfilling them

sbellem opened the pull request:
**IMPORTANT**: targets BigchainDB `0.7.x`

diminator commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+logger = logging.getLogger(__name__)
+
+
+@singledispatch
```
not sure if this adds to readiblity...

diminator commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+        owners_before = list(owners_before)
+
+    if not owners_after:
+        owners_after = owners_before
```
explain?

sbellem replied:
This is to free the user from passing redundant information. 

The way the `CREATE` operation works is that the `owners_before` refers to the original owner(s).  They can make an implicit transfer by setting `owners_after` to different public keys.

The common api requires to pass both `owners_before` and `owners_after` even if they are the same.

diminator replied:
ok - maybe then a line of comment in the code?

sbellem replied:
@diminator Added admonitions `.. important::` e.g.: https://github.com/bigchaindb/bigchaindb-driver/pull/143/files#diff-0de47fef8ba327e59600c5d4d879e81eR157

diminator commented on bigchaindb_driver/utils.py:
```diff
@@ -0,0 +1,41 @@
+
+
+def _normalize_asset(asset, is_transfer=False):
+    if is_transfer:
```
this routine is a bit unclear

From the input arg `asset` you want to make sure you are returning an `Asset` object
One can pass either a `dict` or a list of `kwargs`?
The switch is based on `is_transfer`? Not on type of the arg `asset`?

sbellem replied:
Not sure I fully understand your questions. Here are a few points that may help to clarify:

1. The function is purely for internal use.
2. The function is to help isolate dependencies on `common`.
3. The function is meant to take a `dict` representing an asset, and convert it to a normalized form, which for the time being is an `Asset` instance.

diminator replied:
It is quite tedious to see the difference between

`asset = Asset.from_dict(asset)`
and
`asset = Asset(**asset)`

Whats the structure and type of `asset` in both cases? can you document/comment in the code pls

sbellem replied:
Ok, yes, will do so.

diminator commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+        owners_after = [owners_after]
+
+    fulfillments = [
+        BdbFulfillment(Fulfillment.from_dict(input_['fulfillment']),
```
`Fulfillment.from_dict(input_['fulfillment'])`: why referring to the `fulfillment` of the `input` here?

r-marques replied:
At this point `input_` is actually a condition. So here you are actually creating the fulfillments from the conditions

diminator replied:
I assume that `input_` is a transaction since it has the key `fulfillment`?
really unclear what is happening here

sbellem replied:
yes, agreed, will document the `inputs` parameter, i.e.: give an example

diminator commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+    if not owners_after:
+        owners_after = owners_before
+
+    asset = _normalize_asset(asset)
```
why is there normalization needed?

sbellem replied:
The common api requires an `Asset` instance, whereas the driver apis work strictly with `dict` objects.

So, internally, the `dict` needs to be converted to an `Asset`instance.

This also helps to isolate dependencies on `common`.

diminator commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+        for input_ in inputs
+    ]
+
+    asset = _normalize_asset(asset, is_transfer=True)
```
why is this normalization different?

sbellem replied:
This is due to how the `common` interface works. For `CREATE` operations it is not necessary to supply an asset `id` whereas for `TRANSFER` operations, an asset `id` must be passed. Moreover the way to instantiate an `Asset` object can be slightly different depending on the operation type, hence, the parameter `is_transfer`.

diminator replied:
looking at `_normalize_asset` it seems that the difference if very subtle (haven't figured it out yet)

sbellem replied:
Yes, this is specific to how `common.transaction.Asset` works.

diminator commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+        inputs (:obj:`dict` | :obj:`list` | :obj:`tuple`): One or more
+            inputs holding the condition(s) that this transaction
+            intents to fulfill. Each input is expected to be a
+            :obj:`dict`.
```
structure of the `dict`? `{txid: <ID>, cid: <int>}`?

sbellem replied:
I assume you are suggesting it could be documented? That would be a good point.

It is the complete outer structure. Say you have a `tx` payload you are using as an "input":

```python
cid = 0
condition = tx['transaction']['conditions'][cid]
{
    'fulfillment': condition['condition']['details'],
    'input': {
        'cid': cid,
        'txid': tx['id'],
    },
    'owners_before': condition['owners_after'],
}
```

diminator commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+    fulfilled.
+
+    Args:
+        owners_before (:obj:`list` | :obj:`tuple` | :obj:`str`): One
```
why are there `owners_before`? Isn't this always the/a federation/node?

sbellem replied:
Oh, this was changed. I think the change was made under https://github.com/bigchaindb/bigchaindb/commit/50b0b3cef217bc412dba5a062de1db2d1adcd514 but not sure.

@TimDaub and @r-marques may have more insights on this

r-marques replied:
This changed. Now any client can create assets.
Before we enforced that only federation nodes could create assets but we never had any custom rules for this. We would just accept any create request from clients.
This also takes care of improving the security of create transactions, since the client signs the create transaction no one else can change its content

diminator commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+
+def fulfill_transaction(transaction, *, private_keys):
+    """
+    Fulfills the given transaction.
```
only for multisig?

r-marques replied:
`Transaction.sign` already takes care of both cases

r-marques commented on bigchaindb_driver/driver.py:
```diff
@@ -119,43 +120,91 @@ class TransactionsEndpoint(NamespacedDriver):
+        .. important::
+
+            * For ``operation='CREATE'``, ``owners_before`` MUST be set,
+              meanwhile ``owners_after``, ``asset``, amd ``metadata`` MAY be
```
`amd` -> `and`

r-marques commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+    .. important::
+
+        * For ``operation='CREATE'``, ``owners_before`` MUST be set,
+          meanwhile ``owners_after``, ``asset``, amd ``metadata`` MAY be
```
`amd` -> `and`

r-marques commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+          set. The argument ``inputs`` is ignored.
+
+        * For ``operation='TRANSFER'``, ``owners_after``, ``asset``,
+          and ``inputs`` MUST be set, meanwhile  ``mwtadata`` MAY be set.
```
`mwtadata` -> `metadata`

r-marques commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+
+    Args:
+        owners_before (:obj:`list` | :obj:`tuple` | :obj:`str`): One
+            or more public keys representing the original owner(s) of
```
Instead of `the original owner(s) of` we could use `the issuer of`

r-marques commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+    Returns:
+        dict: The prepared ``"CREATE"`` transaction.
+
+    """
```
Its worth adding and `.. important::` here saying that if `owners_after` is not specified, it will be set to the same as `owners_before`

r-marques commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+    """
+    if not isinstance(inputs, (list, tuple)):
+        inputs = (inputs, )
+    if not isinstance(owners_after, (list, tuple)):
```
won't this have the same problem as https://github.com/bigchaindb/bigchaindb/issues/797?

sbellem replied:
Ah! Since https://github.com/bigchaindb/bigchaindb/issues/797 is not implemented yet. The above was meant to go around the limitation of having to pass lists. 

Did you mean to say that if a `tuple` is passed it will cause troubles?

r-marques replied:
yes

sbellem replied:
yes, you are right I think, maybe too much wishful thinking on my part here -- will review and adjust accordingly

r-marques commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+
+    fulfillments = [
+        BdbFulfillment(Fulfillment.from_dict(input_['fulfillment']),
+                       input_['owners_before'],
```
Shouldn't this be the `owners_after` of the condition?

sbellem replied:
as discussed offline, the field name (key) is called `owners_before` but its value is the `owners_after` from the condition

just to make this clearer for others, say you have a transaction, `tx` you wish to consume:

```python
cid = 0
condition = tx['transaction']['conditions'][cid]
input_ = {
    'fulfillment': condition['condition']['details'],
    'input': {
        'cid': cid,
        'txid': tx['id'],
    },
    'owners_before': condition['owners_after'],
}
```

sbellem commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+
+@_prepare_transaction.register(TransferOperation)
+def _prepare_transfer_transaction_dispatcher(operation, **kwargs):
+    del kwargs['owners_before']
```
Yes, the registered function will receive all the arguments of the generic function (the one decorated with `@singledispatch`). The arguments that are not needed are simply deleted.

diminator replied:
sure, I understand the deletion, but at the time not `owners_before`.
This has been resolved in one of your other comments

r-marques commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,229 @@
+    return transaction.to_dict()
+
+
+def fulfill_transaction(transaction, *, private_keys):
```
Do we really need the `*`? There are no keyword elements in this method

sbellem replied:
As discussed offline, the reason for doing so was that this will allow us to introduce the support for hashlocks and other cryptoconditions goodies in a smooth way.

Hashlocks won't require private keys, and therefore we can add a default value of `None` without breaking current users code.

r-marques commented on bigchaindb_driver/driver.py:
```diff
@@ -183,6 +232,44 @@ def status(self, txid):
         path = self.path + txid + '/status'
         return self.transport.forward_request(method='GET', path=path)
 
+    def create(self, asset=None, verifying_key=None, signing_key=None):
```
Do we still need the `create` and `transfer` methods?

sbellem replied:
As discussed offline, we do not. We will add a deprecation notice for this version and we'll remove it in the subsequent version.

r-marques commented on bigchaindb_driver/utils.py:
```diff
@@ -0,0 +1,41 @@
+            asset = Asset(**asset)
+        except (AttributeError, TypeError):
+            if not asset:
+                asset = None
```
This is ok for the scope of this pull request.
But overall there is still a bit of confusion because some parts of the code try to be isolated from common (argument validation, raising of exceptions) while other parts still rely on common to perform these tasks

sbellem commented:
Note to @diminator and @r-marques: all comments have been addressed except for those that have  `todo` notes (in https://github.com/bigchaindb/bigchaindb-driver/pull/143/commits/87f984fbba00a7d16450641adcb90baae030b5a1)

Will write up the required docs, and will let you know once done.

sohkai commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,325 @@
+        dict: The prepared ``"CREATE"`` transaction.
+
+    .. important:: If ``owners_after`` is not given, or evaluates to
+        ``False``, it will be set equal to ``woners_before``::
```
``owners_before``

sohkai commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,325 @@
+            transaction. Defaults to ``None``.
+
+    Returns:
+        dict: The prepared ``"CREATE"`` transaction.
```
We usually refer to `CREATE` and `TRANSFER` transactions without the quotes (e.g. in the old docs: https://docs.bigchaindb.com/projects/server/en/v0.6.0/drivers-clients/python-server-api-examples.html).

sbellem replied:
The reason is I used the quotes was to keep a connection with the parameter it is linked to, i.e.: `operation`.

The potential user of the library reading the docs, will pass `'CREATE'` or `'TRANSFER'`, with quotes.

That is the reasoning behind the choice to use quotes.

sohkai commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,325 @@
+              'timestamp': '1479393278'},
+             'version': 1}
+
+        Then, the input may be constrcuted in this way::
```
constr**uc**ted

sohkai commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,325 @@
+    Returns:
+        dict: The prepared ``"TRANSFER"`` transaction.
+
+    Example:
```
I would maybe put this in one of the docs files and link to it from here. This might be a little bit excessive for docstrings, especially for people who are going to be changing this file.

sbellem replied:
Yes. Let's see what @diminator says here, because he had asked about an example of input.

sohkai commented on bigchaindb_driver/utils.py:
```diff
@@ -0,0 +1,93 @@
+    """Class representing the ``'TRANSFER'`` transaction operation."""
+
+
+ops_map = {
```
Use an Enum class instead, and just use values from that instead of strings for the two types of transactions. This should eventually land somewhere upstream.

Single dispatch is still possible with [`@valuedispatch`](http://lukasz.langa.pl/8/single-dispatch-generic-functions/#generic-history).

sbellem commented:
@diminator @r-marques done.

Let me know if you have any more comments, feedback, questions.

I have addressed your feedback in new commits, to make it easier to review.

Once you are done, I will rebase the branch to clean up the history.

sohkai commented on bigchaindb_driver/utils.py:
```diff
@@ -0,0 +1,93 @@
+    return operation
+
+
+def _normalize_asset(asset, is_transfer=False):
```
Might as well be `@single_dispatch`ed.

sohkai commented on bigchaindb_driver/driver.py:
```diff
@@ -119,43 +120,100 @@ class TransactionsEndpoint(NamespacedDriver):
+            * ``owners_after``, ``asset``, and ``metadata`` MAY be set.
+            * The argument ``inputs`` is ignored.
+            * If ``owners_after`` is not given, or evaluates to
+              ``False``, it will be set equal to ``woners_before``::
```
``owners_before``

sohkai commented on bigchaindb_driver/offchain.py:
```diff
@@ -0,0 +1,325 @@
+    try:
+        signed_transaction = transaction_obj.sign(private_keys)
+    except KeypairMismatchException as exc:
+        raise MissingSigningKeyError('A signing key is missing!') from exc
```
Maybe we should note this exception in the docstrings both here and upstream in the driver's static method?

sbellem replied:
yes

sbellem replied:
done

sohkai commented on docs/advanced-usage.rst:
```diff
@@ -144,66 +183,146 @@ index ``cid``.
+If you look at the content of ``signed_tx_transfer`` you should see the added
+fulfilment uri, holding the signature:
+
+.. code-block:: python
```
I would just put the `signed_tx_transfer['transaction']['fulfillments'][0]['fulfillment']` block here.

sohkai commented on docs/advanced-usage.rst:
```diff
@@ -144,66 +183,146 @@ index ``cid``.
+    >>> 'cf:4:IMe7QSL5xRAYIlXon76ZonWktR0NI02M8rAG1bN-ugg4S_S7Obu7E-HtL2ZjM3tcKKfoaspMhyx17Eg2KBijylZMxv1NvAD0j8uJP1WOb2AP6ezJorcw6TA5n-cmuwkE'
+
+We have yet to send the transaction over to a BigchainDB node, as both
+preparing and fulfilling a transaction are done "offchain", that is withou the
```
withou**t**

sohkai commented on docs/advanced-usage.rst:
```diff
@@ -256,48 +376,144 @@ Multiple Owners
+     'owners_before': ['3Cxh1eKZk3Wp9KGBWFS7iVde465UvqUKnEqTg2MW4wNf',
+      'EcRawy3Y22eAUSS94vLF8BVJi62wbqbD9iSUSUNU9wAA']}
+
+Them we need the asset:
```
The**n**

sohkai commented on tests/test_driver.py:
```diff
@@ -133,3 +136,41 @@ def test_status_not_found(self, driver):
             driver.transactions.status(txid)
+
+    def test_prepare(self, driver, alice_pubkey):
+        transaction = driver.transactions.prepare(owners_before=[alice_pubkey])
```
Maybe this could be a fixture that can then also be used by `test_fulfill` rather than the hardcoded `unsigned_transaction`?

sbellem replied:
the hard-coding was intentional as it provides a view of what the payload looks like

sohkai commented on tests/test_offchain.py:
```diff
@@ -0,0 +1,91 @@
+        prepare_transaction(operation=None)
+
+
+@mark.parametrize('asset', (
```
Might as well parametrize some `metadata` too?

sohkai commented on tests/test_offchain.py:
```diff
@@ -0,0 +1,91 @@
+    assert 'id' in create_transaction
+    assert 'transaction' in create_transaction
+    assert 'version' in create_transaction
+    assert 'asset' in create_transaction['transaction']
```
Would be nice to actually check the `asset` matched up with the given one. (Or checking assets and metadata can be separate tests)

sohkai commented on tests/test_offchain.py:
```diff
@@ -0,0 +1,91 @@
+    assert 'id' in transfer_transaction
+    assert 'transaction' in transfer_transaction
+    assert 'version' in transfer_transaction
+    assert 'asset' in transfer_transaction['transaction']
```
Would be nice to double check if the asset generated for the transfer matches up with the input's.
